index.js

```diff
@@ -243,6 +246,58 @@ module.exports = function(version, _options) {
             }
 
             return output;
+        },
+        getBestMatchingLanguage: function(language) {
+            if (languages.instructions[language]) return language;
+
+            // If the language is not found, try a little harder
+            var splitLocale = language.toLowerCase().split('-');
+            var languageCode = false;
+            var scriptCode = false;
+            var countryCode = false;
+
+            if (splitLocale.lenth === 1) {
+                languageCode = splitLocale[0];
+            } else if (splitLocale.length === 2 && splitLocale[1].length === 4) {
+                languageCode = splitLocale[0];
+                scriptCode = splitLocale[1];
+            } else if (splitLocale.length === 2 && splitLocale[1].length === 2) {
+                languageCode = splitLocale[0];
+                countryCode = splitLocale[1];
+            } else if (splitLocale.length === 3) {
+                languageCode = splitLocale[0];
+                scriptCode = splitLocale[1];
+                countryCode = splitLocale[2];
+            }
+
+            var supportedLanguageCodes = languages.supportedCodes.map(function(locale) {
+                return locale.toLowerCase().split('-')[0];
+            });
+
+            // Same language code and script code (lng-Scpt)
+            if (languages.instructions[languageCode + '-' + scriptCode]) {
+                return languageCode + '-' + scriptCode;
+
+            // Same language code and country code (lng-CC)
+            } else if (languages.instructions[languageCode + '-' + countryCode]) {
+                return languageCode + '-' + countryCode;
+
+            // Same language code (lng)
+            } else if (languages.instructions[languageCode]) {
+                return languageCode;
+
+            // Same language code and any script code (lng-Scpx) and the found language contains a script
+            } else if (supportedLanguageCodes.indexOf(languageCode) > -1 && scriptCode &&
+                languages.supportedCodes[supportedLanguageCodes.indexOf(languageCode)].split('-').length > 1 &&
+                languages.supportedCodes[supportedLanguageCodes.indexOf(languageCode)].split('-')[1].length === 4) {
```

Comment:
This would be easier if, right off the bat, we map each item in `supportedCodes` to an object with the keys `language`, `script`, and `region`.

Reply:
It'd only be used really for this case. I'm fine with keeping this.

Reply:
If we do the preprocessing suggested above, then this becomes:

```js
else if (availableLanguages.find(function (language) {
  return language.language === languageCode && language.script;
})
```

and some of the duplication in the next `else if` block goes away.

We can do that preprocessing inside languages.js to avoid any meaningful performance penalty.

Reply:
Changed it around to this in https://github.com/Project-OSRM/osrm-text-instructions/pull/168/commits/9a19e1eacde4f7bbdc48543c1f26b49476ce284a, not sure if it's any more clear.